common/big_messaging_typesafe/src/main/scala/uk/ac/wellcome/bigmessaging/typesafe/VHS.scala

```diff
@@ -0,0 +1,124 @@
+package uk.ac.wellcome.bigmessaging.typesafe
+
+import java.util.UUID
+import scala.util.{Failure, Success, Try}
+import org.scanamo.{DynamoFormat, DynamoValue}
+import org.scanamo.error.DynamoReadError
+import org.scanamo.auto._
+import com.typesafe.config.Config
+import com.amazonaws.services.s3.AmazonS3
+import com.amazonaws.services.dynamodbv2.AmazonDynamoDB
+
+import uk.ac.wellcome.typesafe.config.builders.EnrichConfig._
+
+import uk.ac.wellcome.storage.store.s3.S3TypedStore
+import uk.ac.wellcome.storage.store.dynamo.DynamoHashStore
+import uk.ac.wellcome.storage.dynamo.DynamoConfig
+import uk.ac.wellcome.storage.store.s3.S3TypedStore
+import uk.ac.wellcome.storage.store.{
+  HybridIndexedStoreEntry,
+  HybridStoreWithMaxima,
+  Store,
+  TypedStore,
+  VersionedHybridStore
+}
+import uk.ac.wellcome.storage.{
+  Identified,
+  ObjectLocation,
+  ObjectLocationPrefix,
+  Version
+}
+import uk.ac.wellcome.storage.typesafe.{DynamoBuilder, S3Builder}
+import uk.ac.wellcome.storage.streaming.Codec
+import uk.ac.wellcome.storage.maxima.Maxima
+
+case class EmptyMetadata()
+
+trait GetLocation {
+  def getLocation(key: Version[String, Int]): Try[ObjectLocation]
+}
+
+class VHS[T](val hybridStore: VHSInternalStore[T])
+    extends VersionedHybridStore[
+      String,
+      Int,
+      ObjectLocation,
+      T,
+      EmptyMetadata
+    ](hybridStore)
+    with GetLocation {
+
+  def getLocation(key: Version[String, Int]): Try[ObjectLocation] =
+    hybridStore.indexedStore.get(key) match {
+      case Right(Identified(_, entry)) => Success(entry.typedStoreId)
+      case Left(error)                 => Failure(error.e)
+    }
+}
+
+class VHSInternalStore[T](
+  prefix: ObjectLocationPrefix,
+  indexStore: Store[
+    Version[String, Int],
+    HybridIndexedStoreEntry[ObjectLocation, EmptyMetadata]
+  ] with Maxima[String, Int],
+  dataStore: TypedStore[ObjectLocation, T]
+) extends HybridStoreWithMaxima[String, Int, ObjectLocation, T, EmptyMetadata] {
+
+  override val indexedStore = indexStore;
+  override val typedStore = dataStore;
+
+  override protected def createTypeStoreId(
+    id: Version[String, Int]): ObjectLocation =
+    prefix.asLocation(
+      id.id.toString,
+      id.version.toString,
+      UUID.randomUUID().toString)
+}
+
+object VHSBuilder {
+
+  // Scanamo auto derivation cannot derive format for EmptyMetadata correctly.
+  // An underscore is used for variable name as the compiler thinks it is unused
+  // even though it isn't.
+  private implicit val _: DynamoFormat[EmptyMetadata] =
+    new DynamoFormat[EmptyMetadata] {
+      override def read(
+        av: DynamoValue): Either[DynamoReadError, EmptyMetadata] =
+        Right(EmptyMetadata())
+      override def write(t: EmptyMetadata): DynamoValue =
+        DynamoValue.fromMap(Map.empty)
+    }
```

Comment:
My mind spots a similarity between this and https://github.com/wellcometrust/storage-service/blob/f683d8a85c3732a05df5ee2b380eadf4966e0f5c/common/src/main/scala/uk/ac/wellcome/platform/archive/common/bagit/models/ExternalIdentifier.scala#L17-L22

I find that a bit simpler to read, wonder if it's worth a try here? e.g.

```scala
DynamoFormat.iso[DynamoFormat, Map[String, String]](
  _ => EmptyMetadata()
)(
  _ => Map.empty
)
```

Reply:
Tried this:

```
  private implicit val _: DynamoFormat[EmptyMetadata] =
    DynamoFormat.iso[EmptyMetadata, Map[String, String]](
      _ => EmptyMetadata())(
      _ => Map.empty)
```

but resulted in this:

```
11:05:58.307 [default-akka.actor.default-dispatcher-4] ERROR u.a.wellcome.messaging.sqs.SQSStream - Unrecognised failure while: Boxed Error
java.util.concurrent.ExecutionException: Boxed Error
	at scala.concurrent.impl.Promise$.resolver(Promise.scala:87)
	at scala.concurrent.impl.Promise$.scala$concurrent$impl$Promise$$resolveTry(Promise.scala:79)
	at scala.concurrent.impl.Promise$KeptPromise$.apply(Promise.scala:406)
	at scala.concurrent.Promise$.fromTry(Promise.scala:142)
	at scala.concurrent.Future$.fromTry(Future.scala:639)
	at uk.ac.wellcome.platform.recorder.services.RecorderWorkerService.processMessage(RecorderWorkerService.scala:33)
	at uk.ac.wellcome.platform.recorder.services.RecorderWorkerService.$anonfun$run$1(RecorderWorkerService.scala:28)
	at uk.ac.wellcome.bigmessaging.message.MessageStream.$anonfun$foreach$2(MessageStream.scala:57)
	at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:307)
	at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:41)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: java.lang.Error: DynamoReadError: InvalidPropertiesError(NonEmptyList((payload,InvalidPropertiesError(NonEmptyList((metadata,NoPropertyOfType(M,DynNull)))))))
	at uk.ac.wellcome.storage.store.dynamo.DynamoReadable.getEntry(DynamoReadable.scala:29)
	at uk.ac.wellcome.storage.store.dynamo.DynamoReadable.getEntry$(DynamoReadable.scala:23)
	at uk.ac.wellcome.storage.store.dynamo.DynamoHashStore.getEntry(DynamoStore.scala:34)
	at uk.ac.wellcome.storage.store.dynamo.DynamoHashStore.max(DynamoStore.scala:45)
	at uk.ac.wellcome.storage.store.HybridStoreWithMaxima.max(HybridStoreWithMaxima.scala:14)
	at uk.ac.wellcome.storage.store.HybridStoreWithMaxima.max$(HybridStoreWithMaxima.scala:14)
	at uk.ac.wellcome.bigmessaging.typesafe.VHSInternalStore.max(VHS.scala:57)
	at uk.ac.wellcome.storage.store.VersionedStore.getLatest(VersionedStore.scala:56)
	at uk.ac.wellcome.storage.store.VersionedStore.update(VersionedStore.scala:42)
	at uk.ac.wellcome.storage.store.VersionedStore.upsert(VersionedStore.scala:35)
	at uk.ac.wellcome.platform.recorder.services.RecorderWorkerService.storeWork(RecorderWorkerService.scala:45)
	... 11 common frames omitted
```

so keeping as is for now